CLI instrument Uber JAR compresses nested JARs

### Steps to reproduce

* JaCoCo version:  0.8.5
* Operating system: Windows
* Tool integration: Agent / CLI
* Complete executable reproducer: https://github.com/tomcruise81/jacoco-demo.git
* Steps: (what exactly are you doing with the above reproducer?)
    ```bash
    git clone https://github.com/tomcruise81/jacoco-demo.git
    cd jacoco-demo
    mvn package
    targetJar=$(ls target/*.jar | sed 's|target/||')

    # For Windows, under Git Bash, you may also need to convert Cygwin-type paths to Windows paths
    curDir=$(pwd)
    unameOut="$(uname -s)"
    case "${unameOut}" in
        CYGWIN*)    ;&
        MINGW*)     curDir=$(cygpath -m ${curDir});;
    esac

    # Download the appropriate JaCoCo JARs
    curl -k -L https://search.maven.org/remotecontent?filepath=org/jacoco/org.jacoco.cli/0.8.5/org.jacoco.cli-0.8.5-nodeps.jar -o ${curDir}/jacoco-cli.jar
    curl -k -L https://search.maven.org/remotecontent?filepath=org/jacoco/org.jacoco.agent/0.8.5/org.jacoco.agent-0.8.5-runtime.jar -o ${curDir}/jacoco-agent-runtime.jar

    # Run JaCoCo
    java -jar ${curDir}/jacoco-cli.jar instrument ${curDir}/target/${targetJar} --dest ${curDir}/instrumented
    java -javaagent:${curDir}/jacoco-agent-runtime.jar=destfile=${curDir}/jacoco.exec,excludes=*,output=file,dumponexit=true -jar ${curDir}/instrumented/${targetJar}
    ```

### Expected behavior

My expectation is that the pre-instrumented Uber JAR should continue to be runnable and running it with the **javaagent** enabled should allow code coverage output.

I've had to go about it this way instead of just running the main Uber JAR with the **javaagent** due to the output of just running with the **javaagent** being unusable:

```bash
java -javaagent:${curDir}/jacoco-agent-runtime.jar=destfile=${curDir}/jacoco.exec,excludes=*,output=file,dumponexit=true -jar ${curDir}/target/${targetJar}
java -jar ${curDir}/jacoco-cli.jar report jacoco.exec --classfiles=${curDir}/target/${targetJar} --html ${curDir}/coverage-html
```
resulting in:

    [INFO] Loading execution data file C:\test\jacoco-demo\jacoco.exec.
    Exception in thread "main" java.io.IOException: Error while analyzing C:\test\jacoco-demo\target\jacoco-demo-0.0.1-SNAPSHOT.jar@BOOT-INF/lib/log4j-api-2.12.1.jar@org/apache/logging/log4j/util/Base64Util.class.
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzerError(Analyzer.java:162)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeClass(Analyzer.java:134)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeClass(Analyzer.java:157)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeAll(Analyzer.java:193)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeZip(Analyzer.java:265)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeAll(Analyzer.java:196)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeZip(Analyzer.java:265)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeAll(Analyzer.java:196)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeAll(Analyzer.java:226)
            at org.jacoco.cli.internal.commands.Report.analyze(Report.java:110)
            at org.jacoco.cli.internal.commands.Report.execute(Report.java:84)
            at org.jacoco.cli.internal.Main.execute(Main.java:90)
            at org.jacoco.cli.internal.Main.main(Main.java:105)
    Caused by: java.lang.IllegalStateException: Can't add different class with same name: org/apache/logging/log4j/util/Base64Util
            at org.jacoco.cli.internal.core.analysis.CoverageBuilder.visitCoverage(CoverageBuilder.java:106)
            at org.jacoco.cli.internal.core.analysis.Analyzer$1.visitEnd(Analyzer.java:99)
            at org.jacoco.cli.internal.asm.ClassVisitor.visitEnd(ClassVisitor.java:326)
            at org.jacoco.cli.internal.core.internal.flow.ClassProbesAdapter.visitEnd(ClassProbesAdapter.java:100)
            at org.jacoco.cli.internal.asm.ClassReader.accept(ClassReader.java:692)
            at org.jacoco.cli.internal.asm.ClassReader.accept(ClassReader.java:400)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeClass(Analyzer.java:116)
            at org.jacoco.cli.internal.core.analysis.Analyzer.analyzeClass(Analyzer.java:132)
            ... 11 more

### Actual behavior
    java -javaagent:${curDir}/jacoco-agent-runtime.jar=destfile=${curDir}/jacoco.exec,excludes=*,output=file,dumponexit=true -jar ${curDir}/instrumented/${targetJar}
    Exception in thread "main" java.lang.IllegalStateException: Failed to get nested archive for entry BOOT-INF/lib/spring-boot-starter-actuator-2.2.4.RELEASE.jar
            at org.springframework.boot.loader.archive.JarFileArchive.getNestedArchive(JarFileArchive.java:113)
            at org.springframework.boot.loader.archive.JarFileArchive.getNestedArchives(JarFileArchive.java:87)
            at org.springframework.boot.loader.ExecutableArchiveLauncher.getClassPathArchives(ExecutableArchiveLauncher.java:69)
            at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
            at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
    Caused by: java.io.IOException: Unable to open nested jar file 'BOOT-INF/lib/spring-boot-starter-actuator-2.2.4.RELEASE.jar'
            at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:261)
            at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:247)
            at org.springframework.boot.loader.archive.JarFileArchive.getNestedArchive(JarFileArchive.java:109)
            ... 4 more
    Caused by: java.lang.IllegalStateException: Unable to open nested entry 'BOOT-INF/lib/spring-boot-starter-actuator-2.2.4.RELEASE.jar'. It has
    been compressed and nested jar files must be stored without compression. Please check the mechanism used to create your executable jar file
            at org.springframework.boot.loader.jar.JarFile.createJarFileFromFileEntry(JarFile.java:287)
            at org.springframework.boot.loader.jar.JarFile.createJarFileFromEntry(JarFile.java:269)
            at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:258)
            ... 6 more

--- a/org.jacoco.core.test/src/org/jacoco/core/instr/InstrumenterTest.java
+++ b/org.jacoco.core.test/src/org/jacoco/core/instr/InstrumenterTest.java
@@ -26,6 +26,7 @@ import java.io.ObjectOutputStream;
 import java.io.OutputStream;
 import java.io.Serializable;
 import java.util.Arrays;
+import java.util.zip.CRC32;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.GZIPOutputStream;
 import java.util.zip.ZipEntry;
@@ -250,18 +251,38 @@ public class InstrumenterTest {
 	public void testInstrumentAll_Zip() throws IOException {
 		ByteArrayOutputStream buffer = new ByteArrayOutputStream();
 		ZipOutputStream zipout = new ZipOutputStream(buffer);
-		zipout.putNextEntry(new ZipEntry("Test.class"));
+
+		// Compressed Entry
+		ZipEntry entry = new ZipEntry("TestCompressed.class");
+		entry.setMethod(ZipEntry.DEFLATED);
+		zipout.putNextEntry(entry);
 		zipout.write(TargetLoader.getClassDataAsBytes(getClass()));
+
+		// Uncompressed Entry
+		entry = new ZipEntry("TestUncompressed.class");
+		entry.setMethod(ZipEntry.STORED);
+		entry.setSize(TargetLoader.getClassDataAsBytes(getClass()).length);
+		CRC32 crc = new CRC32();
+		crc.update(TargetLoader.getClassDataAsBytes(getClass()));
+		entry.setCrc(crc.getValue());
+		zipout.putNextEntry(entry);
+		zipout.write(TargetLoader.getClassDataAsBytes(getClass()));
+
 		zipout.finish();
 		ByteArrayOutputStream out = new ByteArrayOutputStream();
 
 		int count = instrumenter.instrumentAll(
 				new ByteArrayInputStream(buffer.toByteArray()), out, "Test");
 
-		assertEquals(1, count);
+		assertEquals(2, count);
 		ZipInputStream zipin = new ZipInputStream(
 				new ByteArrayInputStream(out.toByteArray()));
-		assertEquals("Test.class", zipin.getNextEntry().getName());
+		entry = zipin.getNextEntry();
+		assertEquals("TestCompressed.class", entry.getName());
+		assertEquals(ZipEntry.DEFLATED, entry.getMethod());
+		entry = zipin.getNextEntry();
+		assertEquals("TestUncompressed.class", entry.getName());
+		assertEquals(ZipEntry.STORED, entry.getMethod());
 		assertNull(zipin.getNextEntry());
 	}
 

--- a/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
+++ b/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
@@ -229,7 +229,6 @@ public class Instrumenter {
 				final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
 				count += filterOrInstrument(zipin, buffer, name, entryName);
 				final byte[] bytes = buffer.toByteArray();
-				count += filterOrInstrument(zipin, buffer, name, entryName);
 				newEntry.setSize(bytes.length);
 				newEntry.setCompressedSize(bytes.length);
 				final CRC32 crc = new CRC32();

--- a/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
+++ b/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
@@ -16,6 +16,7 @@ import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.util.zip.CRC32;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.GZIPOutputStream;
 import java.util.zip.ZipEntry;
@@ -216,14 +217,40 @@ public class Instrumenter {
 				continue;
 			}
 
-			zipout.putNextEntry(new ZipEntry(entryName));
-			if (!signatureRemover.filterEntry(entryName, zipin, zipout)) {
-				count += instrumentAll(zipin, zipout, name + "@" + entryName);
+			final ZipEntry newEntry = new ZipEntry(entryName);
+			if (entry.getMethod() == ZipEntry.DEFLATED) {
+				newEntry.setMethod(ZipEntry.DEFLATED);
+				zipout.putNextEntry(newEntry);
+				count += filterOrInstrument(zipin, zipout, name, entryName);
+			} else {
+				// Uncompressed entries must be processed in-memory to calculate
+				// mandatory entry size and CRC
+				newEntry.setMethod(ZipEntry.STORED);
+				final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+				count += filterOrInstrument(zipin, buffer, name, entryName);
+				final byte[] bytes = buffer.toByteArray();
+				count += filterOrInstrument(zipin, buffer, name, entryName);
+				newEntry.setSize(bytes.length);
+				newEntry.setCompressedSize(bytes.length);
+				final CRC32 crc = new CRC32();
+				crc.update(bytes);
+				newEntry.setCrc(crc.getValue());
+				zipout.putNextEntry(newEntry);
+				zipout.write(bytes);
 			}
 			zipout.closeEntry();
 		}
 		zipout.finish();
 		return count;
+	}
+
+	private int filterOrInstrument(final InputStream in, final OutputStream out,
+			final String name, final String entryName) throws IOException {
+		if (signatureRemover.filterEntry(entryName, in, out)) {
+			return 0;
+		} else {
+			return instrumentAll(in, out, name + "@" + entryName);
+		}
 	}
 
 	private ZipEntry nextEntry(final ZipInputStream input,

--- a/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
+++ b/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
@@ -232,9 +232,7 @@ public class Instrumenter {
 				final byte[] bytes = buffer.toByteArray();
 				newEntry.setSize(bytes.length);
 				newEntry.setCompressedSize(bytes.length);
-				final CRC32 crc = new CRC32();
-				crc.update(bytes);
-				newEntry.setCrc(crc.getValue());
+				newEntry.setCrc(crc(bytes));
 				zipout.putNextEntry(newEntry);
 				zipout.write(bytes);
 				break;
@@ -254,6 +252,12 @@ public class Instrumenter {
 		} else {
 			return instrumentAll(in, out, name + "@" + entryName);
 		}
+	}
+
+	private static long crc(final byte[] data) {
+		final CRC32 crc = new CRC32();
+		crc.update(data);
+		return crc.getValue();
 	}
 
 	private ZipEntry nextEntry(final ZipInputStream input,

--- a/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
+++ b/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
@@ -218,11 +218,13 @@ public class Instrumenter {
 			}
 
 			final ZipEntry newEntry = new ZipEntry(entryName);
-			if (entry.getMethod() == ZipEntry.DEFLATED) {
+			switch (entry.getMethod()) {
+			case ZipEntry.DEFLATED:
 				newEntry.setMethod(ZipEntry.DEFLATED);
 				zipout.putNextEntry(newEntry);
 				count += filterOrInstrument(zipin, zipout, name, entryName);
-			} else {
+				break;
+			case ZipEntry.STORED:
 				// Uncompressed entries must be processed in-memory to calculate
 				// mandatory entry size and CRC
 				newEntry.setMethod(ZipEntry.STORED);
@@ -236,6 +238,9 @@ public class Instrumenter {
 				newEntry.setCrc(crc.getValue());
 				zipout.putNextEntry(newEntry);
 				zipout.write(bytes);
+				break;
+			default:
+				throw new AssertionError(entry.getMethod());
 			}
 			zipout.closeEntry();
 		}

--- a/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
+++ b/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
@@ -218,16 +218,15 @@ public class Instrumenter {
 			}
 
 			final ZipEntry newEntry = new ZipEntry(entryName);
+			newEntry.setMethod(entry.getMethod());
 			switch (entry.getMethod()) {
 			case ZipEntry.DEFLATED:
-				newEntry.setMethod(ZipEntry.DEFLATED);
 				zipout.putNextEntry(newEntry);
 				count += filterOrInstrument(zipin, zipout, name, entryName);
 				break;
 			case ZipEntry.STORED:
 				// Uncompressed entries must be processed in-memory to calculate
 				// mandatory entry size and CRC
-				newEntry.setMethod(ZipEntry.STORED);
 				final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
 				count += filterOrInstrument(zipin, buffer, name, entryName);
 				final byte[] bytes = buffer.toByteArray();

--- a/org.jacoco.doc/docroot/doc/changes.html
+++ b/org.jacoco.doc/docroot/doc/changes.html
@@ -37,6 +37,13 @@
       (GitHub <a href="https://github.com/jacoco/jacoco/issues/1016">#1016</a>).</li>
 </ul>
 
+<h3>Fixed bugs</h3>
+<ul>
+  <li>Compression method of zip entries is now preserved when instrumenting archives.
+      This allows to use JaCoCo with frameworks that expect uncompressed entries
+      (GitHub <a href="https://github.com/jacoco/jacoco/issues/1004">#1004</a>).</li>
+</ul>
+
 <h3>Non-functional Changes</h3>
 <ul>
   <li>Support for Pack200 was removed in JDK 14. JaCoCo will now throw a detailed

--- a/org.jacoco.doc/docroot/doc/changes.html
+++ b/org.jacoco.doc/docroot/doc/changes.html
@@ -41,7 +41,7 @@
 <ul>
   <li>Compression method of zip entries is now preserved when instrumenting archives.
       This allows to use JaCoCo with frameworks that expect uncompressed entries
-      (GitHub <a href="https://github.com/jacoco/jacoco/issues/1004">#1004</a>).</li>
+      (GitHub <a href="https://github.com/jacoco/jacoco/issues/1018">#1018</a>).</li>
 </ul>
 
 <h3>Non-functional Changes</h3>

